lib/cloudfront/signer.js

```diff
@@ -0,0 +1,182 @@
+     *  @param err [Error]   The error that occurred while calculating the signature
+     *  @param hash [Object] The signature values to set as cookies.
+     */
+    getSignedCookie: function (options, cb) {
+        try {
+            var signatureHash = options.hasOwnProperty('policy')
+                ? signWithCustomPolicy(options.policy, this.keyPairId, this.privateKey)
+                : signWithCannedPolicy(options.url, options.expires, this.keyPairId, this.privateKey);
+
+            var cookieHash = {};
+            for (var key in signatureHash) {
+                if (signatureHash.hasOwnProperty(key)) {
+                    cookieHash['CloudFront-' + key] = signatureHash[key];
+                }
+            }
+
+            cb(null, cookieHash);
+        } catch (err) {
+            cb(err);
+        }
+    },
+
+    /**
+     * Create a signed Amazon CloudFront URL.
+     *
+     * Keep in mind that URLs meant for use in media/flash players may have
+     * different requirements for URL formats (e.g. some require that the
+     * extension be removed, some require the file name to be prefixed
+     * - mp4:<path>, some require you to add "/cfx/st" into your URL).
+     *
+     * @param options [Object]          The options to create a signed URL.
+     * @option options url [String]     The URL to which the signature will grant
+     *                                  access. Required.
+     * @option options expires [Number] A Unix UTC timestamp indicating when the
+     *                                  signature should expire. Required unless you
+     *                                  pass in a full policy.
+     * @option options policy [String]  A CloudFront JSON policy. Required unless
+     *                                  you pass in a url and an expiry time.
+     * @callback cb function(err, url)
+     *  Called when an error occurs or the signature has been calculated.
+     *  @param err [Error]  The error that occurred while calculating the signature
+     *  @param url [String] The signed URL.
+     */
+    getSignedUrl: function (options, cb) {
+        try {
+            var resource = getResource(options.url),
+                parsedUrl = url.parse(options.url, true),
+                signatureHash = options.hasOwnProperty('policy')
+                    ? signWithCustomPolicy(options.policy, this.keyPairId, this.privateKey)
+                    : signWithCannedPolicy(resource, options.expires, this.keyPairId, this.privateKey);
+
+            parsedUrl.search = null;
+            for (var key in signatureHash) {
+                if (signatureHash.hasOwnProperty(key)) {
+                    parsedUrl.query[key] = signatureHash[key];
+                }
+            }
+
+            var signedUrl = url.format(parsedUrl);
+            if (determineScheme(options.url) === 'rtmp') {
+                cb(null, getRtmpUrl(signedUrl));
+            } else {
+                cb(null, signedUrl);
```

Comment:
Here and in getSignedCookie, we should probably check that a callback was supplied before calling it, and return the signedUrl/cookieHash immediately if one wasn't provided.  This is how s3.getSignedUrl works, for example.